File: LeaflektMap/Sources/LeaflektMap/Models/LeaflektMapProperties.swift
import Foundation

public struct LeaflektMapProperties {
    public let mapStyle: LeaflektMapStyle
    public let geoJsonOverlay: LeaflektGeoJsonOverlay
    public let tileBufferSize: Int

    public init(
        mapStyle: LeaflektMapStyle = .openStreetMap,
        geoJsonOverlay: LeaflektGeoJsonOverlay = .india,
        tileBufferSize: Int = {
            let ramMb = Int(ProcessInfo.processInfo.physicalMemory / (1024 * 1024))
            if ramMb < 2048 { return 4 }
            if ramMb < 4096 { return 8 }
            return 12
        }()
    ) {
        self.mapStyle = mapStyle
        self.geoJsonOverlay = geoJsonOverlay
        self.tileBufferSize = tileBufferSize
    }
}

File: leaflekt-core/src/commonMain/kotlin/com/binayshaw7777/leaflekt/LeaflektMapProperties.kt
package com.binayshaw7777.leaflekt

val DefaultLeaflektMapProperties: LeaflektMapProperties = LeaflektMapProperties()

data class LeaflektClusterConfig(
    val groupId: String,
    val maxClusterRadius: Int = 80
)

class LeaflektMapProperties(
    val mapStyle: LeaflektMapStyle = LeaflektMapStyle.OpenStreetMap,
    val geoJsonOverlay: LeaflektGeoJsonOverlay = LeaflektGeoJsonOverlay.India,
    val tileBufferSize: Int = defaultTileBufferSize(),
    val clusterConfig: LeaflektClusterConfig? = null
) {
    override fun toString(): String =
        "LeaflektMapProperties(mapStyle=$mapStyle, geoJsonOverlay=$geoJsonOverlay, tileBufferSize=$tileBufferSize, clusterConfig=$clusterConfig)"

    override fun equals(other: Any?): Boolean =
        other is LeaflektMapProperties &&
            mapStyle == other.mapStyle &&
            geoJsonOverlay == other.geoJsonOverlay &&
            tileBufferSize == other.tileBufferSize &&
            clusterConfig == other.clusterConfig

    override fun hashCode(): Int {
        var result = 31 * (31 * mapStyle.hashCode() + geoJsonOverlay.hashCode()) + tileBufferSize
        result = 31 * result + (clusterConfig?.hashCode() ?: 0)
        return result
    }

    fun copy(
        mapStyle: LeaflektMapStyle = this.mapStyle,
        geoJsonOverlay: LeaflektGeoJsonOverlay = this.geoJsonOverlay,
        tileBufferSize: Int = this.tileBufferSize,
        clusterConfig: LeaflektClusterConfig? = this.clusterConfig
    ): LeaflektMapProperties = LeaflektMapProperties(
        mapStyle = mapStyle,
        geoJsonOverlay = geoJsonOverlay,
        tileBufferSize = tileBufferSize,
        clusterConfig = clusterConfig
    )
}
